Statement logging empty tables (#20711)
MaterializeInc/materialize: doc/user/content/sql/system-catalog/mz_internal.md

| `id`                | [`text`]         | The ID of the source. Corresponds to [`mz_catalog.mz_sources.id`](../mz_catalog#mz_sources).                   |
| `replication_slot`  | [`text`]         | The name of the replication slot in the PostgreSQL database that Materialize will create and stream data from. |

> <!-- ### `mz_prepared_statement_history` -->

Reviewer: Nit: Are multi-line comments possible? If so I suggest putting all of this into a single comment, for a much better diff when you make this a non-comment later.

<!-- The `mz_prepared_statement_history` table contains a subset of all -->
<!-- statements that have been prepared. It only contains statements that -->
<!-- have one or more corresponding executions in inclusion in -->
<!-- [`mz_statement_execution_history`](#mz_statement_execution_history). -->

<!-- | Field         | Type                         | Meaning                                                                                                                           | -->
<!-- |---------------|------------------------------|-----------------------------------------------------------------------------------------------------------------------------------| -->
<!-- | `id`          | [`uuid`]                     | The globally ID of the prepared statement.                                                                                        | -->
<!-- | `session_id`  | [`uuid`]                     | The globally unique ID of the session that prepared the statement. Corresponds to [`mz_session_history.id`](#mz_session_history). | -->

> <!-- | `name`        | [`text`]                     | The name of the prepared statement (the default prepared statement's name is the empty string).                                   | -->

Author: Similar to `prepared_at`, I think it's useful to keep this just in case it can be useful for debugging.

<!-- | `sql`         | [`text`]                     | The SQL text of the prepared statement.                                                                                           | -->

> <!-- | `prepared_at` | [`timestamp with time zone`] | The time at which the statement was prepared.                                                                                     | -->

Reviewer 2: I'm not sure this has any value, because we don't even resolve object names on the stored statement, so users could prepare something that references an object that doesn't yet exist.

Author: You're saying you don't think the `prepared_at` timestamp has value? Maybe not, but I'd rather keep it in case we decide we want it, as it's only a few additional bytes.

<!-- ### `mz_session_history` -->

<!-- The `mz_session_history` table contains all the sessions that have -->
<!-- been established in the last 30 days, or (even if older) that are -->
<!-- referenced from -->
<!-- [`mz_prepared_statement_history`](#mz_prepared_statement_history). -->

<!-- | Field          | Type                         | Meaning                                                                                                                           | -->
<!-- |----------------|------------------------------|-----------------------------------------------------------------------------------------------------------------------------------| -->
<!-- | `id`           | [`uuid`]                     | The globally unique ID of this history entry. Does **not** correspond to [`mz_sessions.id`](#mz_sessions), which can be recycled. | -->
<!-- | `connected_at` | [`timestamp with time zone`] | The time at which the session was established.                                                                                    | -->

> <!-- | `application`  | [`text`]                     | The `application` session metadata field.                                                                                         | -->

Author: > If so, can you include details on how we handle a user changing this value mid-session?

That is a good question which I hadn't considered. I assumed this value was set during session startup and was immutable, but now that I'm looking at the code it seems that was wrong, and it's controlled by a var.

The main reason we want this is to distinguish between different surfaces: psql, mz, the web console, the web shell, etc. Do you know if those are likely to be available on session startup, or do they only get set later? I can also check this myself if you don't know.

> <!-- | `application`  | [`text`]                     | The `application` session metadata field.                                                                                         | -->

Reviewer 3: This is a recurring unsolved issue. The application name is a session variable. Like all session variables it has the following properties: 

- It can be initialized via the connection string.
  - If it's not initialized via the connection string, then it's initialized to some default value, in the case of `application_name` the default value is the empty string.
- It can be modified at any time via `SET application_name ...`.
- Changing the value is done transitionally, i.e. if the value is set in a transaction and the transaction aborts, then the change is rolled back.
- The value can be changed temporarily for an ongoing transaction. i.e. `SET LOCAL application_name ...` will change the application name for the duration of the transaction, then roll it back when the transaction commits.

Given all of this, it's unclear how `application_name` or any other session variable should be written down in a catalog table.

One idea was that we could write down the application name that appeared in the connection string, if any, and indicate in the docs that it may not be the current value of the application name. I think this got shot down, but I can't remember why.

> <!-- | `application`  | [`text`]                     | The `application` session metadata field.                                                                                         | -->

Reviewer 2: ```suggestion
<!-- | `application_name`  | [`text`]                     | The `application_name` session metadata field.                                                                                         | -->
```

> <!-- | `application`  | [`text`]                     | The `application` session metadata field.                                                                                         | -->

Reviewer 2: This is really annoying because some drivers might run this as their first command, but that's after startup has occurred. Regardless, we should start with the value after startup (which will reflect anything in the connection string, including our web console). If that ends up not being enough we can do more.

> <!-- | `application`  | [`text`]                     | The `application` session metadata field.                                                                                         | -->

Author: I think being able to distinguish between psql, mz, web console, and web shell will give us 90% of the possible value here.

> <!-- | `session_user` | [`text`]                     | The name of the user for wish the session was established.                                                                        | -->

Author: Will change it to `authenticated_user`, sure.

As for denormalization: this is a "history" table, in which case I think it makes sense to denormalize, because we are trying to record the state of the world at the time the event happened. Similar to `mz_audit_events`.

<!-- | Field                   | Type                         | Meaning                                                                                                                                                                                                                        | -->
<!-- |-------------------------|------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------| -->
<!-- | `id`                    | [`uuid`]                     | The ID of the execution event.                                                                                                                                                                                                 | -->
<!-- | `prepared_statement_id` | [`uuid`]                     | The ID of the prepared statement being executed. Corresponds to [`mz_prepared_statement_history.id`](#mz_prepared_statement_history).                                                                                          | -->

> <!-- | `sample_rate`           | [`double precision`]         | The sampling rate at the time the execution began.                                                                                                                                                                             | -->

Reviewer 4: So this is the value of the `SystemVar` at the specific point in time?

Tying this back to the discussion above, could a similar approach be a possible solution to the `application_name` session var (I understand that this has implications on the memory requirements)?

<!-- | `params`                | [`text list`]                | The values of the prepared statement's parameters.                                                                                                                                                                             | -->
<!-- | `began_at`              | [`timestamp with time zone`] | The time at which execution began.                                                                                                                                                                                             | -->

> <!-- | `finished_at`           | [`timestamp with time zone`] | The time at which execution ended.                                                                                                                                                                                             | -->

Reviewer 2: What do you imagine this being? For non-peek statements, an `ExecuteResponse` is ready and can immediately be returned to a user. For `SELECT`s, though, a future is returned and resolved only when clusterd has returned the vec of rows, but that is handled much deeper in the row sending code. See https://github.com/MaterializeInc/materialize/pull/20532 for more discussion about this metric.

What about `SUBSCRIBE`? This could measure how long it took to register the subscribe with clusterd.

What about `FETCH`, which contains the name of another portal that needs to be polled for results, and which may need to begin execution itself?

Author: In the version of statement logging that I have working locally, which I will PR in the next 1-2 days or so, the answers are as follows:

* For statements that finish immediately in the coordinator, we are done when we call `ExecuteContext::retire`.
* For `SELECT` statements, we are done when the row future resolves, or immediately in the case of constants that are resolved in the coordinator without a peek ever being sent to a cluster.
* For `SUBSCRIBE` statements, we are done when the subscribe ends (either because it actually finished, which is possible if e.g. subscribing to a constant value, or because the user cancelled it)
* For `FETCH` statements, we are done whenever the inner execute finishes.

The references in [this doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/reference/adapter/coord_interface.md) to when a statement is considered finished are intended to correspond to how they will be treated as part of statement logging.

Reviewer 4: > For `SUBSCRIBE` statements, we are done when the subscribe ends (either because it actually finished, which is possible if e.g. subscribing to a constant value, or because the user cancelled it).

Are we going to emit sampled `SUBSCRIBE` with `finished_at` set to `NULL` until they are done or wait for them to finish before we log them?

*Edit*: I guess having two kinds of entries for "execution begin" vs "execution end" modeled as separate structs means that we're leaning towards the former. I remember from the design doc that we want to do this internally, what I'm not sure is whether this internal state will be exposed as the current state of the corresponding `mz_internal` table.

Author: Yes, the publicly-exposed `finished_at` will be `NULL` while the query is still running.

> <!-- | `was_successful`        | [`bool`]                     | Whether the statement executed successfully. Mutually exclusive with `was_canceled` and `was_aborted`.                                                                                                                         | -->

Reviewer 2: Do we also need an error state?

Author: That is implicit in whether the error message is `NULL` or not. I agree that that's a bit confusing, so I'll add an `e` state when I switch this to using a char-based enum.

<!-- | `was_canceled`          | [`bool`]                     | Whether the statement execution was canceled. Mutually exclusive with `was_successful` and `was_aborted`.                                                                                                                      | -->

> <!-- p| `was_aborted`           | [`bool`]                     | Whether the statement execution was aborted. "Aborted" means that the execution had not finished when Materialize restarted (for maintenance or due to an error). Mutually exclusive with `was_successful` and `was_canceled`. | -->

Author: That's a good idea, will do. Is that a common pattern for expressing "enums" in SQL?

> <!-- p| `was_aborted`           | [`bool`]                     | Whether the statement execution was aborted. "Aborted" means that the execution had not finished when Materialize restarted (for maintenance or due to an error). Mutually exclusive with `was_successful` and `was_canceled`. | -->

Reviewer 3: Anecdotally it feels like a common pattern that PostgreSQL catalog tables use. There exists an enum type (https://www.postgresql.org/docs/current/datatype-enum.html), but I've never seen it used in a catalog table.

> <!-- p| `was_aborted`           | [`bool`]                     | Whether the statement execution was aborted. "Aborted" means that the execution had not finished when Materialize restarted (for maintenance or due to an error). Mutually exclusive with `was_successful` and `was_canceled`. | -->

Reviewer 2: How are we able to detect abortions if they deal with the system restarting.

Author: On startup, we re-read this collection from the stash (currently. We will probably switch to using persist directly rather than the stash). At that time we fill in "aborted" for any non-finished value.

> <!-- | `error_message`         | [`test`]                     | The error returned when executing the statement, or `NULL` if it was successful, canceled or aborted.                                                                                                                          | -->

Reviewer: ```suggestion
<!-- | `error_message`         | [`text`]                     | The error returned when executing the statement, or `NULL` if it was successful, canceled or aborted.                                                                                                                          | -->
```

> <!-- | `rows_returned`         | [`uint8`]                    | The number of rows returned by the statement, if it finished successfully and was of a kind of statement that can return rows, or `NULL` otherwise.                                                                            | -->

Reviewer 2: Users (including us) don't have a reliable way to get uint8s out of our system (there's no binary encoding that's well understood by drivers). Users could cast it to a postgres `int8`, but it's a bit silly to force that anytime this is queried. Recommend making this an `int8`.

> <!-- | `was_fast_path`         | [`bool`]                     | Whether the statement could be answered without building a dataflow (e.g., purely by reading an in-memory index), if it finished successfully and was of a kind of statement that can return rows, or `NULL` otherwise.        | -->

Reviewer 2: There's also a third peek mode for constant expressions. This should maybe be a char field similar to the execution status above:

```suggestion
<!-- | `execution_mode`         | [`char`]                     | The execution mode for rows-returning queries. `'f'` (`f`ast path) if the statement could be answered without building a new dataflow (e.g., purely by reading an in-memory index), `'s'` (`s`low path) if a new dataflow was built, `'c'` (`c`onstant) if the statement was a constant expression and no dataflow was needed, or `NULL` if this was not a rows-returning query.        | -->
```
